### PR TITLE
refactor: code organization

### PR DESCRIPTION
**Pull request by varunsrin**
## Motivation

The Hub has become more complicated and harder to reason about as the size of the codebase as grown. A high level design that we all agree on will make it much easier to read and write code as we are getting ready for the first release.

## Change Summary

The PR is very large, touches many files and squeezed in all the changes I could over a single weekend to reduce merge conflicts during the workweek. A series of smaller followup changes will be made over the next week to bring the rest in line. Comments and suggestions are welcome for other things that should be updated, or patterns that people disagree with. 

The major changes are: 

- Added new README.md that describes the proposed architecture
- Reorganized code under the high level folders `network`, `storage`, `types`, `utils` and `urls`
- Refactored `rpc` into a module, which is a pattern we should replicate over the codebase to keep files small
- Moved non-unit tests files and helpers into `test` folder
- Added documentation to `hubs.ts`
- Enforced stricter type checking rules to minimize type errors

## Merge Checklist

- [x] The title of this PR adheres to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard
- [x] The PR has been tagged with the appropriate change type label(s) (i.e. documentation, feature, bugfix, or chore)
- [ ] Changes to the protocol specification have been merged

## Additional Context

If this is a relatively large or complex change, provide more details here that will help reviewers

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -12,7 +12,7 @@ module.exports = {
     '@typescript-eslint/ban-ts-ignore': 'off',
     '@typescript-eslint/explicit-function-return-type': 'off',
     '@typescript-eslint/no-explicit-any': 'off',
-    'no-restricted-imports': ['error', { patterns: ['.*'] }],
+    'no-restricted-imports': ['error', { patterns: ['"*/lib/*"'] }],
```
Comment (varunsrin) [inline .eslintrc.cjs:15]: note to self: should be `build` not lib

File: README.md [hunk 2]
```diff
@@ -1,18 +1,125 @@
...
+Clients[FC Clients] & Clients & Clients -.-> |RPC| Hub
+Hub <-.-> |RPC + Gossip|Hubs[FC Hubs] & Hubs & Hubs
+Storage <--> Networking
+```
+
+### Storage Engine
+
+The Storage Engine validates the signature and schema of every received message. It computes the set of valid signers for a user by tracking Farcaster contract events and uses this to verify each message.
+
+Verified messages are merged into a Set, which is a data structure represents the current state of the user's data on the network. A Set uses CRDT's to ensure that merging a message is commutative, associative and idempotent. This means that two Hubs that receive a set of messages in any order will always end up in the same state.
```
Comment (sagar-a16z) [inline README.md:55]: nit: which is a data structure `that` represents the

File: src/hub.ts [hunk 3]
```diff
@@ -11,39 +11,44 @@ import {
...
+  /** Address of the IdRegistry contract  */
+  IDRegistryAddress?: string;
+
   /*
-   * Whether simple sync should be used once the network is established
-   * This would typically only be disabled on the first node to join a new network
+   * Enable SimpleSync once network is established.
+   *
+   * @remarks
+   * Usually disabled after the first node joins the network
```
Comment (sagar-a16z) [inline src/hub.ts:40]: Correction: Usually disabled _for_ the first node in a network.

File: src/hub.ts [hunk 4]
```diff
@@ -85,102 +87,81 @@ export class Hub extends TypedEmitter<HubEvents> implements RPCHandler {
...
+  get rpcAddress() {
+    // Safety: RPC is always configured on an IP socket, so it can be cast safely here
+    return this.rpcServer.address ? (this.rpcServer.address as AddressInfo) : undefined;
+  }
+
+  get gossipAddresses() {
+    return this.gossipNode.multiaddrs ?? [];
+  }
+
+  // Returns the Gossip peerId string of this Hub
```
Comment (sagar-a16z) [inline src/hub.ts:105]: Update to this for consistency?
```
/** Returns the Gossip peerId string of this Hub */ 
```

File: src/network/rpc/utils.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import { Result } from 'neverthrow';
```
Comment (sagar-a16z) [inline src/network/rpc/utils.ts:1]: nit: I don't think `utils` is the right name for this. It's the core definition of the RPC API implemented by RPC Servers. 

Can just call it `api`, `interfaces` or something along those lines. The `replacer` and `reviver` here are utils though but the interface declarations here are more important.

**Review by sagar-a16z — APPROVED** [review]

File: src/hub.ts (hunk 4 above)
Comment (varunsrin) [reply src/hub.ts:105]: good catch

File: src/network/rpc/utils.ts (hunk 5 above)
Comment (varunsrin) [reply src/network/rpc/utils.ts:1]: yeah, agreed

**Pull request merged**